lib/src/importer/node_package.dart

```diff
@@ -0,0 +1,398 @@
+          var replaced = string.replaceAll(RegExp(r'\*'), patternMatch);
+          var path = p.normalize(p.join(packageRoot, replaced));
+          return fileExists(path) ? path : null;
+        }
+        return p.join(packageRoot, string);
+      case Map<String, dynamic> map:
+        var conditions = ['sass', 'style', 'default'];
+        for (var (key, value) in map.pairs) {
+          if (!conditions.contains(key)) continue;
+          if (_packageTargetResolve(
+                  subpath, value as Object, packageRoot, patternMatch)
+              case var result?) {
+            return result;
+          }
+        }
+        return null;
+
+      case []:
+        return null;
+
+      case List<dynamic> array:
+        for (var value in array) {
+          var result = _packageTargetResolve(
+              subpath, value as Object, packageRoot, patternMatch);
+          if (result != null) {
+            return result;
+          }
+        }
+
+        return null;
+
+      default:
+        throw "Invalid 'exports' value in ${p.join(packageRoot, 'package.json')}";
+    }
+  }
+
+  /// Returns a path to a package's export without a subpath.
+  Object? _getMainExport(Object exports) {
+    Object? parseMap(Map<String, dynamic> map) {
+      if (!map.keys.any((key) => key.startsWith('.'))) {
+        return map;
+      }
+      if (map.containsKey('.')) {
+        return map['.'] as Object;
+      }
+      return null;
+    }
+
+    return switch (exports) {
+      String string => string,
+      List<String> list => list,
+      Map<String, dynamic> map => parseMap(map),
+      _ => null
+    };
+  }
+
+  /// Returns a list of all possible variations of `subpath` with extensions and
+  /// partials.
+  ///
+  /// `subpath` is part of the [bare import specifier], so we use the `/`
+  /// separator.
+  List<String?> _exportsToCheck(String? subpath, {bool addIndex = false}) {
+    var paths = <String>[];
+
+    if (subpath == null && addIndex) {
+      subpath = 'index';
+    } else if (subpath != null && addIndex) {
+      subpath = p.posix.join(subpath, 'index');
+    }
+    if (subpath == null) return [null];
+
+    if (['scss', 'sass', 'css'].any((ext) => subpath!.endsWith(ext))) {
+      paths.add(subpath);
+    } else {
+      paths.addAll([
+        '$subpath.scss',
+        '$subpath.sass',
+        '$subpath.css',
+      ]);
+    }
+    var subpathSegments = p.posix.split(subpath);
+    var basename = subpathSegments.last;
+    if (!basename.startsWith('_')) {
+      List<String> prefixedPaths = [];
+      for (final path in paths) {
+        var dirPath = subpathSegments.sublist(0, subpathSegments.length - 1);
+        var pathBasename = p.posix.split(path).last;
+        if (dirPath.isEmpty) {
+          prefixedPaths.add('_$pathBasename');
+        } else {
+          dirPath.add('_$pathBasename');
+          prefixedPaths.add(p.posix.joinAll(dirPath));
+        }
+      }
+      paths.addAll(prefixedPaths);
```

Comment:
Rather than modifying a mutable list, try using list comprehensions to construct the return value, as in:

```dart
var dirname = p.dirname(subpath);
return [
  ...paths,
  for (var path in paths)
    if (dirname == '.') '_${p.basename(path)}'
    else p.join(dirname, '_${p.basename(path)}')
];
```